### PR TITLE
feat: canvas replay

### PR DESCRIPTION
**Pull request by daibhin**
## Problem

Related to https://github.com/PostHog/posthog-js/pull/946 and https://github.com/PostHog/posthog/issues/14555

Provide config via decide & playback canvas mutation events

## Changes

- Implement plugin to replace canvas with decoded image from event
- Add new `session_recording_config` to store all future settings related to replay in a single bag
- Add feature flag to prevent setting up the feature unless enabled
- Add settings to enable / disable canvas recording with warning about PII

## How did you test this code?

- Wrote team & decide tests

https://github.com/PostHog/posthog/assets/6685876/8e4b138d-69df-472d-8701-a92d4baf2753

### CODE REVIEW

File: frontend/src/scenes/session-recordings/player/rrweb/canvas/canvas-plugin.ts [hunk 1]
```diff
@@ -0,0 +1,114 @@
+import { CanvasArg, type canvasMutationData, type canvasMutationParam, eventWithTime } from '@rrweb/types'
+import { EventType, IncrementalSource, Replayer } from 'rrweb'
+// TODO: figure out how to import this method
+import canvasMutation from 'rrweb/es/rrweb/packages/rrweb/src/replay/canvas'
```
Comment (daibhin) [inline frontend/src/scenes/session-recordings/player/rrweb/canvas/canvas-plugin.ts:4]: Hoping to get this in via https://github.com/rrweb-io/rrweb/pull/1383

File: posthog/api/test/__snapshots__/test_decide.ambr [hunk 2]
```diff
@@ -89,22 +87,6 @@
   LIMIT 21 /*controller='team-detail',route='api/projects/%28%3FP%3Cid%3E%5B%5E/.%5D%2B%29/%3F%24'*/
   '
 ---
-# name: TestDecide.test_decide_doesnt_error_out_when_database_is_down.10
```
Comment (marandaneto) [inline posthog/api/test/__snapshots__/test_decide.ambr:92]: Nice cleaning up

File: posthog/api/team.py [hunk 3]
```diff
@@ -208,6 +210,18 @@ def validate_session_recording_network_payload_capture_config(self, value) -> Di
...
 
+    def validate_session_recording_config(self, value) -> Dict | None:
+        if value is None:
+            return None
+
+        if not isinstance(value, Dict):
+            raise exceptions.ValidationError("Must provide a dictionary or None.")
+
+        if not all(key in ["record_canvas"] for key in value.keys()):
+            raise exceptions.ValidationError("Must provide a dictionary with only 'record_canvas' key.")
```
Comment (marandaneto) [inline posthog/api/team.py:221]: Maybe it should not be a dictionary then?

File: frontend/src/scenes/session-recordings/player/sessionRecordingPlayerLogic.ts [hunk 4]
```diff
@@ -526,6 +527,10 @@ export const sessionRecordingPlayerLogic = kea<sessionRecordingPlayerLogicType>(
                 plugins.push(CorsPlugin)
             }
 
+            if (values.featureFlags[FEATURE_FLAGS.SESSION_REPLAY_CANVAS]) {
```
Comment (pauldambra) [inline frontend/src/scenes/session-recordings/player/sessionRecordingPlayerLogic.ts:530]: is it possible to get in a state where I'm collecting canvas recordings but this flag is off?

not blocking here but we should have a good story for "this replay has a canvas in it"

Comment (pauldambra) [reply frontend/src/scenes/session-recordings/player/sessionRecordingPlayerLogic.ts:530]: also makes me think it'd be good to have a PostHog insight showing web vs android vs canvas vs etc playback so we can measure feature usage

File: posthog/api/team.py (hunk 3 above)
Comment (pauldambra) [reply posthog/api/team.py:221]: Ah, I'd forgotten my pre-xmas good citizen push that we should move session replay config into one place so we can stop adding more nad more columns to the team model

Maybe we should get this in and then I use it for the config in https://github.com/PostHog/posthog/pull/19887

File: posthog/models/team/team.py [hunk 5]
```diff
@@ -184,6 +184,7 @@ class Team(UUIDClassicModel):
     )
     session_recording_linked_flag: models.JSONField = models.JSONField(null=True, blank=True)
     session_recording_network_payload_capture_config: models.JSONField = models.JSONField(null=True, blank=True)
+    session_recording_config: models.JSONField = models.JSONField(null=True, blank=True)
```
Comment (pauldambra) [inline posthog/models/team/team.py:187]: nitpicking... should it be `session_replay_config`? (am open to be convinced not :))

**Review by pauldambra — APPROVED** [review]
👍 I can validate that it doesn't break existing recordings :)

How do I test canvas recordings? (maybe being silly)

File: posthog/models/team/team.py (hunk 5 above)
Comment (daibhin) [reply posthog/models/team/team.py:187]: Nope, think you're right

**Comment by daibhin** [comment]
@pauldambra it's a little tricky / awkward. I added a `/canvas` route to the NextJS playgorund app in posthog-js. I run that app with `NEXT_PUBLIC_POSTHOG_KEY=<% ph_local_key %> NEXT_PUBLIC_POSTHOG_HOST=http://localhost:8000 yarn dev` to point at my local instance.

I also have an exported recording that I can send you 😄

File: frontend/src/scenes/session-recordings/player/sessionRecordingPlayerLogic.ts (hunk 4 above)
Comment (daibhin) [reply frontend/src/scenes/session-recordings/player/sessionRecordingPlayerLogic.ts:530]: Yes, possibly could be. We should also consider in future that recordings might be captured with canvas and then the config option is turned off. Should we continue to playback recordings with the canvas or not in that situation 🤔 

Good news is that it's pretty easy to know if a recording contains a canvas in the rrweb plugin during playback

**Pull request merged**